### PR TITLE
Add support for new annotation and OLM descriptor format

### PR DESCRIPTION
**Pull request by isutton**
### Motivation

Refer to https://github.com/k8s-service-bindings/spec#annotation-examples for the format.

### Changes

This pull request substitutes the annotations' syntax with those specified in https://github.com/k8s-service-bindings/spec#annotation-examples.

The following use cases have been covered:

#### Contribute an entire Secret data to the binding

The annotation indicates that `.status.data.dbCredentials` in the Service resource must exist and be the name of a Secret in the same namespace.

```yaml
service.binding: path={.status.data.dbCredentials},objectType=Secret
```

The following is present in the Service resource, the contents of `dbCredentials` being a string with the name of a Secret existing in the same namespace as the Service itself.

```yaml
status:
  data:
    dbCredentials: database-cred-Secret  # Secret
```

The following content is present in the `database-cred-Secret` Secret resource:

```yaml
data:
  username: <base64 encoded string>
  password: <base64 encoded string>
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `DBCREDENTIALS_USERNAME` | decoded string |
| `DBCREDENTIALS_PASSWORD` | decoded string |

---

#### Contribute an entire ConfigMap data to the binding

The annotation indicates that `.status.data.dbConfiguration` in the Service resource must exist and be the name of a ConfigMap in the same namespace.

```yaml
service.binding: path={.status.data.dbConfiguration},objectType=ConfigMap
```

The following is present in the Service resource, the contents of `dbConfiguration` being a string with the name of a ConfigMap existing in the same namespace as the Service itself.

```yaml
status:
  data:
    dbConfiguration: database-config     # ConfigMap
```

The following content is present in the `database-config` ConfigMap resource:

```yaml
data:
  port: 5432
  ipAddress: 192.168.1.250
  certificate: <base64 encoded certificate>
  db_timeout: 30
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `DBCONFIGURATION_PORT` | `5432` |
| `DBCONFIGURATION_IPADDRESS` | `192.168.1.250` |
| `DBCONFIGURATION_CERTIFICATE` | `<base64 encoded certificate>` |
| `DBCONFIGURATION_DB_TIMEOUT` | `30` |

---

#### Contribute a single entry from a ConfigMap

The annotation indicates that `.status.data.dbConfiguration` in the Service resource must exist and be the name of a ConfigMap in the same namespace, and the field `certificate` will be contributed under the `certificate` name.

```yaml
service.binding/certificate: path={.status.data.dbConfiguration},objectType=ConfigMap,sourceKey=certificate
```

The following is present in the Service resource, the contents of `dbConfiguration` being a string with the name of a ConfigMap existing in the same namespace as the Service itself.

```yaml
status:
  data:
    dbConfiguration: database-config     # ConfigMap
```

The following content is present in the `database-config` ConfigMap resource:

```yaml
data:
  port: 5432
  ipAddress: 192.168.1.250
  certificate: <base64 encoded certificate>
  db_timeout: 30
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `CERTIFICATE` | `<base64 encoded certificate>` |

---

#### Contribute a single entry from a ConfigMap under a different name

The annotation indicates that `.status.data.dbConfiguration` in the Service resource must exist and be the name of a ConfigMap in the same namespace, and the field `db_timeout` will be contributed under the `timeout` name.

```yaml
service.binding/timeout: path={.status.data.dbConfiguration},objectType=ConfigMap,sourceKey=db_timeout
```

The following is present in the Service resource, the contents of `dbConfiguration` being a string with the name of a ConfigMap existing in the same namespace as the Service itself.

```yaml
status:
  data:
    dbConfiguration: database-config     # ConfigMap
```

The following content is present in the `database-config` ConfigMap resource:

```yaml
data:
  port: 5432
  ipAddress: 192.168.1.250
  certificate: <base64 encoded certificate>
  db_timeout: 30
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `TIMEOUT` | `30` |

---

#### Contribute a single entry from the Service resource

The annotation indicates that `.status.data.url` in the Service resource is going to be contributed as the `uri` name.

```yaml
service.binding/uri: path={.status.data.url}
```
The following is present in the Service resource:

```yaml
status:
  data:
    url: db.stage.ibm.com
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `URI` | `db.stage.ibm.com` |

---

#### Contribute a slice of maps from a Service resource field

The annotation indicates that a slice of maps resulting in a `map[string]string` built extracting the `type` key as key and the `url` key as value; it expects the value in `.status.bootstrap` to be a `[]map[string]interface{}` and available as `endpoints`:

```yaml
service.binding/endpoints: path={.status.bootstrap},elementType=sliceOfMaps,sourceKey=type,sourceValue=url
```

The following content is present in the Service resource:

```yaml
status:
  bootstrap:
  - type: plain
    url: myhost2.example.com
    name: hostGroup1
  - type: tls
    url: myhost1.example.com:9092,myhost2.example.com:9092
    name: hostGroup2
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `ENDPOINTS_PLAIN` | `myhost2.example.com` |
| `ENDPOINTS_TLS` | `myhost1.example.com:9092,myhost2.example.com:9092` |

---

#### Contribute a slice of strings from a Service resource field

The annotation indicates that a `[]string` is built extracting the `type` key as key and the `url` key as value; it expects the value in `.status.bootstrap` to be a `[]map[string]interface{}` and available as `endpoints`:

```yaml
service.binding/bootstrap: path={.status.bootstrap},elementType=sliceOfStrings,sourceValue=url
```

The following content is present in the Service resource:

```yaml
status:
  bootstrap:
  - type: plain
    url: myhost2.example.com
    name: hostGroup1
  - type: tls
    url: myhost1.example.com:9092,myhost2.example.com:9092
    name: hostGroup2
```

Once the `ServiceBindingRequest` is reconciled, the following service binding values should be exposed to the application as environment variables:

| Variable | Value |
|---|---|
| `BOOTSTRAP_0` | `myhost2.example.com` |
| `BOOTSTRAP_1` | `myhost1.example.com:9092,myhost2.example.com:9092` |

### Testing

WIP

For further more details refer the CONTRIBUTING.md

### CODE REVIEW

**Comment by sbose78** [comment]
Is this support for the new annotations in the spec?

**Comment by baijum** [comment]
> Is this support for the new annotations in the spec?

I would suggest change the title and description to mention the new "annotation format". Something like this:
```
Support annotation format recommended by the spec

This PR adds support for the annotation format recommended by the
k8s-service-bindings spec (https://github.com/k8s-service-bindings/spec#annotation-examples).
```

**Comment by sbose78** [comment]
+1

File: docs/BackingServiceBestPractices.md [hunk 1]
```diff
@@ -1,169 +1,286 @@
...
+4. #### Use “certificate” from the ConfigMap “status.data.dbConfiguration” as a volume mount
+
+    Requirement : *Extract a specific field from the configmap/Secret from the Kubernetes resource and use it as a volume mount.*
+
+
+    Annotation
+
+    ```
+    “servicebinding.dev/certificate”:
+    "path={.status.data.dbConfiguration},bindAs=volume,objectType=ConfigMap"
```
Comment (sbose78) [inline docs/BackingServiceBestPractices.md:136]: Is `bindAs=volume` being supported in this PR?

Comment (pedjak) [reply docs/BackingServiceBestPractices.md:136]: no.

**Pull request merged**